packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -22,31 +22,42 @@ class OverlayEntry {
   bool get opaque => _opaque;
   bool _opaque;
   void set opaque (bool value) {
+    assert(_overlay != null);
     if (_opaque == value)
       return;
-    _opaque = value;
-    markNeedsBuild();
+    _overlay.setState(() {
+      _opaque = value;
+    });
   }
 
-  OverlayState _state;
+  OverlayState _overlay;
+  final GlobalKey _key = new GlobalKey();
 
   /// Remove this entry from the overlay.
   void remove() {
-    _state?._remove(this);
-    _state = null;
+    _overlay?._remove(this);
+    _overlay = null;
   }
 
   /// Cause this entry to rebuild during the next pipeline flush.
   ///
   /// You need to call this function if the output of [builder] has changed.
   void markNeedsBuild() {
-    // TODO(ianh): find a way to make this not rebuild the entire overlay
-    _state?.setState(() {});
+    _key.currentState?.setState(() { /* the state that changed is in the builder */ });
   }
 
   String toString() => '$runtimeType@$hashCode(opaque: $opaque)';
 }
 
+class _OverlayEntry extends StatefulComponent {
```

Comment:
I considered it, but I couldn't figure out a clean way of getting the setState out of it. I mean, you could capture it, but that seems like an ugly precedent to set. This seemed like a cleaner solution.